### DISCUSSION TITLE
HA not working properly

### DISCUSSION THREAD

**Discussion by lyandr** [requarks/wiki#3274] [Error / Bug Report]
**Question**
Hello, i have deployed Wikijs on Openshift. HA_ACTIVE=1, using one node first and after settings upgrade to two nodes.

When i create a new page on master i can see this on slave after F5 and vice versa.

But when i edit a page on master, the update is not apply on slave and vice versa even after F5.

Did i missed something on my configuration? How to be sure that HA is well activated? Or is there any limitations due to the fact that file are locally rendered in container because if i go on the opposite node and i rerendered all pages the content is updated.

**Host Info (please complete the following information):**
OS: Linux - Alpine Linux (linux) 3.11.6 x64
Wiki.js version: 2.5.170
Database engine: postgres 9.6.10

**Comment by lyandr** [comment]
I saw in the logs that HA is active when container is starting:
`2020-12-07T13:26:37.337Z [MASTER] info: High-Availability Listener initialized successfully: [ OK ]`

But i don't see any trigger event in the log as its seems to be : 
`WIKI.logger.info(`Received event ${payload.event} from instance ${payload.source}: [ OK ]`)`

When i update a page content i only see the following log on the concerned container:

```
Loading configuration from /wiki/config.yml...
--
  | OK
  | 2020-12-09T15:19:25.866Z [JOB] info: Rendering page ID 3...
  | 2020-12-09T15:19:27.209Z [JOB] info: Using database driver pg for postgres [ OK ]
  | 2020-12-09T15:19:30.129Z [JOB] info: Rendering page ID 3: [ COMPLETED ]
```

**Comment by lyandr** [comment]
is there any one implements HA and have this issue?

**Comment by lyandr** [comment]
Hello is it possible to have some informations about HA and my problem?

**Comment by haydenseitz** [comment]
I'm also having troubles with HA.

I have a docker swarm stack configured to run 3 replicas of the app. During the install (when you enter admin password), I set the replica count to 1, and once the install is complete and I get the options Create Home Page or Administration, I then set the replica count back to 3.

I see the log message about the HA process starting:
`2022-03-31T17:32:19.265Z [MASTER] info: High-Availability Listener initialized successfully: [ OK ]`

But when I'm logged in as the admin and changing settings in the administration section, I get "unauthorized" banner warnings. It only works on every 3rd reload, when I assume it makes the request to the instance that I first logged in at.

Can any of the code owners comment on how the replication works between instances? Is there a discovery process to find other instances? is it multicast? Thanks!

**Reply by NGPixel** [reply]
Are all your instances pointing to the same DB? Everything is done via the DB. Events that need to be broadcasted to other instances are done via PostgreSQL pub/sub protocol.

**Reply by vuducmanh11** [reply]
I'm using wikijs cluster 3 instrances connect VIP patroni postgres exposed by haproxy and have same issue.

**Comment by lilianabiven** [comment]
Also having troubles with HA, pages not replicating properly when edited on one node. Just checking if there are any plans to improve it in the next major version? Thanks!

**Comment by vuducmanh11** [comment]
I have same issue. 3 wikijs instance connect postgres patroni cluster via haproxy VIP

**Comment by NGPixel** [comment]
HA will be significantly improved in v3. You have to understand that HA was patched on after the fact in v2 so there are times where it can get out of sync or not properly receive cache purge commands. v3 is built from the ground up with HA in mind.

In the meantime, restart your instances if you notice problems, as that should resolve pretty much all issues you may experience.

**Comment by cmseal** [comment]
For anyone else that finds this discussion - I've seen that in NodeJS v12.x, the pub/sub works without issue (tested with WikiJS v2.5.285).  Keeping that version of WikiJS but installing NodeJS >12 causes it to no longer work (tried on 14.x and 16.x, then 20.x with WikiJS v2.5.300).

I'm not suggesting using NodeJS v12.x given how old and unsupported etc., but the difference seems to be related to after v12.x.

**Comment by p64** [comment]
I definitely had the same problem.  The admin interface seems to ping pong.  I have two instances balanced.  If I update a field and save it, it comes back original.  Then a refresh, every other time it's updated.  I have it load balanced with standard nginx.  Waiting for V3 seems best at this point.

Product is great!  not trying to dig on of the developers.  Thanks for a great product.

**Comment by angelod-cr** [comment]
Had the same problem. Updating the **pg-pubsub** package version from _0.5.0_ to _0.8.1_ did the trick for me

**Comment by akire0ne** [comment]
It seems that we have the same issues on our 3 replicas HA setup. 
Users complains that if they refresh multiple times a freshly edited page, they will get the old and the new version intermittently. 
Restarting instances seems to do the trick but it is not a viable solution.
